test/unit_tests/cvode/CXX_serial/cv_test_resize_history.cpp

```diff
@@ -0,0 +1,423 @@
+ * Programmer(s): David J. Gardner @ LLNL
+ * -----------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2025, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------------------
+ * Test resizing the CVODE history array
+ * ---------------------------------------------------------------------------*/
+
+#include <cstdio>
+#include <iomanip>
+#include <iostream>
+
+#include "cvode/cvode.h"
+#include "cvode/cvode_impl.h"
+#include "nvector/nvector_serial.h"
+#include "sundials/sundials_context.hpp"
+#include "sundials/sundials_linearsolver.h"
+#include "sundials/sundials_matrix.h"
+#include "sundials/sundials_nvector.h"
+#include "sunlinsol/sunlinsol_band.h"
+#include "sunmatrix/sunmatrix_band.h"
+#include "sunnonlinsol/sunnonlinsol_fixedpoint.h"
+
+#include "problems/pr.hpp"
+#include "utilities/check_return.hpp"
+
+using namespace std;
+using namespace problems::pr;
+
+constexpr sunrealtype zero = SUN_RCONST(0.0);
+constexpr sunrealtype one  = SUN_RCONST(1.0);
+
+static int save_history(sunrealtype t_n, N_Vector y_n, sunrealtype* t_hist,
+                        N_Vector* y_hist, N_Vector* f_hist, int hist_size,
+                        int step, void* user_data)
+{
+  // Move old values back one overwriting the oldest value and insert the most
+  // recent time, solution vector, and rhs vector as the first history entry
+  int i_start = hist_size - 2;
+  if (step < hist_size) { i_start = step - 1; }
+
+  for (int i = i_start; i >= 0; i--) { t_hist[i + 1] = t_hist[i]; }
+  t_hist[0] = t_n;
+
+  for (int i = i_start; i >= 0; i--)
+  {
+    N_VScale(one, y_hist[i], y_hist[i + 1]);
+    N_VScale(one, f_hist[i], f_hist[i + 1]);
+  }
+  N_VScale(one, y_n, y_hist[0]);
+  int retval = ode_rhs(t_n, y_n, f_hist[0], user_data);
+  if (retval) { return 1; }
+
+  return 0;
+}
+
+static int resize_history(sunrealtype* t_hist, N_Vector* y_hist, N_Vector* f_hist,
+                          int hist_size, SUNContext sunctx, void* user_data)
+{
+  // Add one new element to the solution and rhs history vectors
+  int new_size = N_VGetLength(y_hist[0]) + 1;
+
+  for (int i = 0; i < hist_size; i++)
+  {
+    N_Vector new_vec      = N_VNew_Serial(new_size, sunctx);
+    sunrealtype* old_data = N_VGetArrayPointer(y_hist[i]);
+    sunrealtype* new_data = N_VGetArrayPointer(new_vec);
+    for (int j = 0; j < new_size; j++) { new_data[j] = old_data[0]; }
+    N_VDestroy(y_hist[i]);
+    y_hist[i] = new_vec;
+  }
+
+  for (int i = 0; i < hist_size; i++)
+  {
+    N_VDestroy(f_hist[i]);
+    f_hist[i]  = N_VClone(y_hist[i]);
+    int retval = ode_rhs(t_hist[i], y_hist[i], f_hist[i], user_data);
+    if (retval) { return 1; }
+  }
+
+  return 0;
+}
```

Comment:
I can also imagine a CVODE utility routine that would handle something like this as well, by taking a user-supplied function to "project" a vector of the old size onto a vector of the new size, or perhaps two projection functions: one for `y` and the other for `f`?

Reply:
In a case where CVODE saves the necessary history, yes having a function (or functions) like this would be nice way to fill the history vectors. For the current use case, saving the history in CVODE is unnecessary as the user code already has save the full history and this would require additional memory.